---
create: 2023-07-08
modified: '2024-11-17'
---

## 1.2 抛出异常

#### 1.书本内容

如下是一个抛出异常的例子：

```C++
//: C01:MyError.cpp {RunByHand}
class MyError
{
    const char *const data;
public:
    MyError(const char *const msg = 0) : data(msg) {}
};

void f()
{
    // Here we "throw" an exception object:
    throw MyError("something bad happened");
}

int main()
{
    // As you'll see shortly,we'll want a "try block" here:
    f();
}

```

注意：

* 异常返回地点与return返回地点不一样。
* 局部对象会被正常销毁
* 抛出不同类型的异常以帮助判断

---

#### 2.额外内容

> 注：由于在抛出异常时，如果没有故意做好准备，那么一般调用的是拷贝构造函数。

##### 在抛出异常时，如何使用移动构造函数优化异常对象的创建：

假设我们有一个类**MyException**，它继承自**std::exception**，并且有一个成员变量**msg**来存储异常信息。我们可以为它定义一个拷贝构造函数和一个移动构造函数，如下：

```c++
class MyException : public std::exception {
public:
    // 普通构造函数
    MyException(const char* msg) : msg(msg) {}

    // 拷贝构造函数
    MyException(const MyException& other) : msg(other.msg) {
        std::cout << "Copy constructor of MyException\n";
    }

    // 移动构造函数
    MyException(MyException&& other) noexcept : msg(std::move(other.msg)) {
        std::cout << "Move constructor of MyException\n";
    }

private:
    std::string msg;
};
```

在这个例子中，我们可以看到：

- 拷贝构造函数会复制原对象的msg字符串，并且输出一条信息表示调用了拷贝构造函数。
- 移动构造函数会移动原对象的msg字符串，并且输出一条信息表示调用了移动构造函数。注意我们使用了noexcept关键字来声明它不会抛出异常。
- 如果我们在某个函数中需要抛出一个MyException对象，比如：

```c++
void foo() {
    throw MyException("Something wrong");
}
```

那么编译器会优先选择使用移动构造函数来创建异常对象，因为它更高效，并且不会抛出异常。如果我们**没有定义移动构造函数**，或者**没有声明它为noexcept**，那么编译器会使用拷贝构造函数来代替它。